src/mahalanobis.jl

```diff
@@ -1,19 +1,19 @@
 # Mahalanobis distances
 
-struct Mahalanobis{T} <: Metric
-    qmat::Matrix{T}
+struct Mahalanobis{T,M<:AbstractMatrix{T}} <: Metric
+    qmat::M
 end
 
-struct SqMahalanobis{T} <: SemiMetric
-    qmat::Matrix{T}
+struct SqMahalanobis{T,M<:AbstractMatrix{T}} <: SemiMetric
+    qmat::M
 end
 
-result_type(::Mahalanobis{T}, ::Type, ::Type) where {T} = T
-result_type(::SqMahalanobis{T}, ::Type, ::Type) where {T} = T
+result_type(::Mahalanobis{T,M}, ::Type, ::Type) where {T,M} = T
+result_type(::SqMahalanobis{T,M}, ::Type, ::Type) where {T,M} = T
 
 # SqMahalanobis
 
-function (dist::SqMahalanobis{T})(a::AbstractVector, b::AbstractVector) where {T <: Real}
+function (dist::SqMahalanobis)(a::AbstractVector, b::AbstractVector)
```

Comment:
Nice, I like that we are removing the restriction on Real here as well.

Reply:
In fact, these restrictions are "wrong", because they don't apply to the type, right? Formally, we don't have distance computation methods for eltypes other than `<:Real`.

Reply:
Yep. I think we can eliminate the restriction in the code, and let users figure out if a particular implementation actually works with other numbers that aren't real. I think this is one of those antipatterns in Julia where we prematurely restrict the type without a clear benefit.

Reply:
echo 👍 I like this change

I was quite annoyed with these specific type annotations because the whole JuliaImages world are built on top of `Gray{Float32}` things :P and none of them are Real...

```julia
julia> Gray{Float32} <: Colorant
true

julia> Gray{Float32} <: Real
false
```